x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/MergedGeoLines.java

```diff
@@ -45,131 +44,225 @@ public double[] getFinalSortValues() {
+            return getLeftChildIndex(i) < size;
+        }
 
-    private boolean hasRightChild(int i) {
-        return getRightChildIndex(i) < size;
-    }
+        private boolean hasRightChild(int i) {
+            return getRightChildIndex(i) < size;
+        }
+
+        private void heapifyUp() {
+            int i = size - 1;
+            while (hasParent(i) && correctOrdering(getParentIndex(i), i)) {
+                int parentIndex = getParentIndex(i);
+                swap(parentIndex, i);
+                i = parentIndex;
+            }
+        }
+
+        private void heapifyDown() {
+            int i = 0;
+            while (hasLeftChild(i)) {
+                int childIndex = getLeftChildIndex(i);
+                if (hasRightChild(i) && correctOrdering(getRightChildIndex(i), childIndex) == false) {
+                    childIndex = getRightChildIndex(i);
+                }
+                if (correctOrdering(childIndex, i)) {
+                    break;
+                } else {
+                    swap(childIndex, i);
+                    i = childIndex;
+                }
+            }
+        }
 
-    private void heapifyUp() {
-        int i = size - 1;
-        while (hasParent(i) && correctOrdering(getParentIndex(i), i)) {
-            int parentIndex = getParentIndex(i);
-            swap(parentIndex, i);
-            i = parentIndex;
+        private void swap(int i, int j) {
+            int tmpLineIndex = lineIndices[i];
+            int tmpIdxWithinLine = idxsWithinLine[i];
+            lineIndices[i] = lineIndices[j];
+            idxsWithinLine[i] = idxsWithinLine[j];
+            lineIndices[j] = tmpLineIndex;
+            idxsWithinLine[j] = tmpIdxWithinLine;
         }
     }
 
-    private void heapifyDown() {
-        int i = 0;
-        while (hasLeftChild(i)) {
-            int childIndex = getLeftChildIndex(i);
-            if (hasRightChild(i) && correctOrdering(getRightChildIndex(i), childIndex) == false) {
-                childIndex = getRightChildIndex(i);
+    /**
+     * This version assumes all InternalGeoLines passed into the class are not overlapping
+     * and as such we do not need to perform a merge-sort, and instead can simply concatenate ordered lines.
+     * The final result is either truncated or simplified based on the simplify parameter.
+     */
+    static final class NonOverlapping extends MergedGeoLines {
+
+        NonOverlapping(List<InternalGeoLine> geoLines, int finalLength, SortOrder sortOrder, boolean simplify) {
+            super(geoLines, finalLength, sortOrder, simplify);
+        }
+
+        /**
+         * merges <code>geoLines</code> into one sorted list of values representing the combined line.
+         */
+        @Override
+        public void merge() {
+            TreeSet<InternalGeoLine> sorted = switch (sortOrder) {
+                case DESC -> new TreeSet<>((o1, o2) -> Double.compare(o2.sortVals()[0], o1.sortVals()[0]));
```

Comment:
Maybe use static comparators? And also `naturalOrdering` and `reverseOrder` comparators.